### PR TITLE
Amendments: how to tell if rippled is amendment blocked

### PR DESCRIPTION
**Pull request by jhaaaa**
* Adds a few tips for verifying that your `rippled` server is amendment blocked.
* Adds a tip for finding out which amendments are causing the block.
* Will add a bit more details about how and why a `rippled` server becomes amendment blocked.

### CODE REVIEW

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: It's probably fine to put it here for now.

Long-term, I think this'll fit into section VI.3 of the rippled doc plan ("The rippled Server - Advanced Configuration")

Nitpick: capitalize "is" or don't capitalize "if". I think the former is proper style, although I'm no expert.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Is this change necessary? Rules can "apply", active voice, can't they?

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Good question. I'd like a developer to confirm.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Add a caution that versions older than 0.80.0 (per RIPD-1479?) don't include this field even if your server is amendment blocked.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Respectively:

1. Yes.
2. Yes. (Confirmed [in the code](https://github.com/ripple/rippled/blob/3e9e0c4b53aa29775b02f5a092c79587cebef8d0/src/ripple/app/misc/NetworkOPs.cpp#L2141-L2142)—server_info and server_state use the same underlying function and there's no special code to change how this field displays.)
3. Yes.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Yes, with a link back to this page.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: The typical case is you're running a version of `rippled` that's older than the one used by trusted validators.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: They shouldn't. Even if they do, that's a separate task.

**Review by mDuo13 — CHANGES_REQUESTED** [review]
Answered the questions I could, picked some nits, looks good overall.

If you're having trouble actually amendment-blocking your server, that's not surprising. The code for (more explicitly) amendment-blocking your server is new in 0.80.0 and the newest amendment currently enabled on the production XRP Ledger is from 0.80.0. Your best bet might be to connect a v0.80.0 server to the test net where some 0.90.0 amendments are already turned on.

File: content/concept-amendments.md
Comment (ryangyoung) [reply content/concept-amendments.md]: I think the question might be what's applying the rules to what? My interpretation is that In the original, the rules are the agent applying themselves to the ledger. In the amended version, `rippled` has applied the rules to the ledger. FWIW, the amended version makes more intuitive sense to me as a (maybe not very smart) reader.

File: content/concept-amendments.md
Comment (ryangyoung) [reply content/concept-amendments.md]: You always want to upgrade to latest version, right? Would there be a case where you _don't_ want to use a certain amendment? (I can't think of why you'd want to do that, just asking.)

**Review by ryangyoung — COMMENTED** [review]
LGTM.

File: content/concept-amendments.md
Comment (mDuo13) [reply content/concept-amendments.md]: It's certainly not guaranteed that you would want to use a certain amendment. Amendments can change _anything_ about the ledger's transaction processing, so the changes could be distasteful or unprofitable for you. Alternatively, you may find a bug in an amendment's code.

If the rest of the network approves an amendment, your only choices are to accept that and move on, or fork away from the network from a point before the amendment became enabled. Something akin to the latter is what created the Bitcoin Cash and Ethereum Classic forks. (Bitcoin and Ethereum protocols don't have a formalized amendment process in the way the XRP Ledger does, but that doesn't stop them from changing their rules. It just means that changes to their behavior are far less predictable and servers that aren't upgraded automatically fork off rather than becoming amendment blocked).

Comment (ryangyoung) [reply content/concept-amendments.md]: So that means it sort of doesn't matter _which_ amendments aren't supported, right? If _any_ of them are `supported: false`, you have to upgrade to the latest `rippled` version. Else, the decision point described. 

I guess I'm pointing this out because the way the amendments are specified now sort of makes it seem like you could pick and choose which ones you want to support (emphasis on "sort of"). That is, in a way it doesn't matter which specific ones are causing the block; you've got to take all of them if you want to get unblocked.

File: content/concept-amendments.md
Comment (ryangyoung) [reply content/concept-amendments.md]: Here's what the _Microsoft Manual of Style says about title capitalization:

>* Capitalize all nouns, verbs (including is and other forms of be), adverbs (including than and when), adjectives (including this and that), and pronouns (including its).
>* Capitalize the first and last words, regardless of their parts of speech (“The Text to Look For”).
>*  Capitalize prepositions that are part of a verb phrase (“Backing Up Your Disk”).
>*  Do not capitalize articles (a, an, the) unless the article is the  first word in the title.
>*  Do not capitalize coordinate conjunctions (and, but, for, nor, or, yet, so) unless the conjunction is the first word in the title.
>*  Do not capitalize prepositions of four or fewer letters unless the preposition is the first or the last word in the title.
>*  Do not capitalize to in an infinitive phrase (“How to Format Your Hard Disk”) unless the phrase is the first word in the title.
>*  Capitalize the second part of a hyphenated compound if it would be capitalized without the hyphen. Always capitalize the second part of a hyphenated compound if it is the last word of a heading or title.

File: content/concept-amendments.md
Comment (mDuo13) [reply content/concept-amendments.md]: If any amendments are both enabled _and_ unsupported, you must upgrade or fork.

For amendments that are unsupported and _not_ enabled, you can "veto" them (configure your server not to vote in favor of them) and hope that other trusted validators do the same. You're not amendment blocked until those amendments become enabled.

File: content/concept-amendments.md
Comment (jhaaaa) [reply content/concept-amendments.md]: I am no expert either, but from what I've read, 'if" is a subordinating conjunction and should be capitalized -- per AP and Chicago, for example. Sound okay? What is the next style guide in line when Microsoft Manual of Style doesn't specify something?

Comment (ryangyoung) [reply content/concept-amendments.md]: The IBM Style Guide. AP and Chicago deviate can pretty far from tech writing standards. AP's stance on serial commas, for instance: yuck.

Comment (mDuo13) [reply content/concept-amendments.md]: The Microsoft one is more specific than most other style guides I saw when it comes to title case. It seems like most style guides—contrary to what I would've expected—suggest you should capitalize "Is" though.

Comment (jhaaaa) [reply content/concept-amendments.md]: IBM Style Guide on subordinating conjunctions: In headline-style capitalization, capitalize the initial letter of the following words ... subordinating conjunctions, such as after, although, because, before, how, if, than, that, though, until, when, where, whether, and while.

File: content/concept-amendments.md
Comment (jhaaaa) [reply content/concept-amendments.md]: @mDuo13 and @ryangyoung - Yes, thank you for this thread! 🙏  This is exactly what I was trying to suss out in trying to amendment block my server -- and then I was going to see what it takes to get it out of that state. 

Got it - so if you are amendment blocked, it may be interesting to know which features are blocking you -- but regardless, you cannot cherry pick which amendments ("enabled" and "unsupported") you want to take to unblock yourself. You just need to upgrade to the version of `rippled` that will unblock you. 

And just to be sure that I fully understand -- this may not always be the latest version of `rippled`, is that right? Meaning, can there be a `rippled` version that does not amendment block previous versions? If this is true, you could have a scenario in which upgrading to a version before the latest version can unblock you. Do we think there are folks operating servers who would have a use case for needing to know how to do this? Meaning, per the doc, the requirement would be to upgrade to the version that will unblock you (I can write a doc on how to determine this), but the recommendation would be to just upgrade to the latest version. 

For amendments that are "not enabled" and "unsupported", you do still have a choice in terms of being able to veto specific amendments, but this discussion is not about getting out of an amendment blocked state, but rather is important in the context of voting, correct?

File: content/concept-amendments.md
Comment (jhaaaa) [reply content/concept-amendments.md]: @mDuo13 - Checked with Brandon W - he's thinking that this is not a universal error: It is currently only returned for the subset of commands that are designated with `NEEDS_CURRENT_LEDGER` and  `NEEDS_CLOSED_LEDGER`. https://github.com/ripple/rippled/blob/c3a6b1600fffd5e7c22d790b29e09fe0e09fcfff/src/ripple/rpc/impl/RPCHandler.cpp#L160-L165
https://github.com/ripple/rippled/blob/1c44c4a43eb9e46d8270892a9e49466ec7cc3efb/src/ripple/rpc/impl/Handler.cpp#L100

File: content/concept-amendments.md
Comment (mDuo13) [reply content/concept-amendments.md]: > Can there be a rippled version that does not amendment block previous versions?

Yes, absolutely. I mean, new versions themselves don't automatically amendment-block previous versions. Different versions of `rippled` can coexist as long as no enabled amendments are only known by some versions.

> Do we think there are folks operating servers who would have a use case for needing to know how to [upgrade to not the newest version]?

Some people like to thoroughly read the code or wait for others to test a new version.  Those people may end up a version behind everyone else, which is probably OK depending on the version and why it was released. (If it has security fixes, they should really upgrade ASAP.)

> Meaning, per the doc, the requirement would be to upgrade to the version that will unblock you (I can write a doc on how to determine this), but the recommendation would be to just upgrade to the latest version.

Yes.

> For amendments that are "not enabled" and "unsupported", you do still have a choice in terms of being able to veto specific amendments, but this discussion is not about getting out of an amendment blocked state, but rather is important in the context of voting, correct?

Yes. This is described in the [amendment voting documentation](https://ripple.com/build/amendments/#amendment-voting). To "veto" an amendment means to configure your server not to vote in favor of that amendment. (Anyone not voting in favor of an amendment is effectively voting against it.)

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: > `rippled` error

I suggest the phrase "log message" instead since that indicates where you'd see this message.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Yeah, I was thinking about changing the Known Amendments table to list them like that. It quickly starts to run off the side of the page on some screens, though. That's something to address in a different task.

File: content/rippled-api-methods/server_info.md
Comment (mDuo13) [inline content/rippled-api-methods/server_info.md]: Add a _(May be omitted)_ marker, and rephrase to match the "closed_ledger" phrasing a little more closely, and use a badge for the new feature. Namely:

> If `true`, this server is [amendment blocked] (concept-amendments.html#amendment-blocked). If the server is not amendment blocked, the response omits this field. [New in: rippled 0.80.0][] |

(You'll also have to add 0.80.0 to `snippets/rippled_versions.md`)

File: content/rippled-api-methods/server_state.md
Comment (mDuo13) [inline content/rippled-api-methods/server_state.md]: Edit to match the recommendations from server_info

**Review by mDuo13 — CHANGES_REQUESTED** [review]
Left a couple more suggestions, very minor at this point.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: After re-reading this in context, I think I would suggest changing "backend applications" to "applications that depend on XRP Ledger data". My rationale is that "backend" is a fairly vaguely defined term which people tend to use in varying ways, and also because JavaScript web-apps (typically considered "frontend" code) can connect to `rippled` directly via WebSocket, but amendment blocking is meant to protect them, too.

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: > A server only becomes amendment blocked if the enabled amendment

I suggest changing this to "if _an_ enabled amendment"

(as usual, apologies for asking you to fix things that you didn't break)

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Make "when you submit a transaction" a link to the submit command documentation. (`reference-rippled.html#submit`)

Also, that reminds me, the submit command docs need "amendmentBlocked" in the "Possible Errors" table. Can you cover that in this PR, too? 🙇

File: content/concept-amendments.md
Comment (mDuo13) [inline content/concept-amendments.md]: Don't forget to remove this TODO before merging.

**Review by mDuo13 — CHANGES_REQUESTED** [review]
Found one missing piece—the `amendmentBlocked` error response in the submit command reference docs—and one "TODO" still there. Other than that, looks fantastic.

File: content/concept-amendments.md
Comment (jhaaaa) [reply content/concept-amendments.md]: 🙏

**Review by mDuo13 — APPROVED** [review]
_and_ you sorted the possible errors from submit? 💯

**Pull request merged**
